### ISSUE TITLE
Create open boundary conditions and initial config.

### ISSUE THREAD

**Issue by AndyHoggANU** [COSIMA/mom6-eac#1]
Use same "cookie-cutter" strategy as `mom6-panan` to create open boundary conditions from a single RYF year of ACCESS-OM2.

For topography, we should begin with `ACCESS-OM2-01` topography, before pushing to 1/20°using etopo data.

**Comment by angus-g** [comment]
Thanks for getting this started! Do you have any suggestions for the forcing dataset? I guess we can start with daily boundary forcing like we have already, is there a year of daily T, S, U, V (and maybe eta)?

**Comment by AndyHoggANU** [comment]
I think for the first iteration we should do exactly as we did for mom6-panan.
That is, the surface forcing is JRA-55-RYF9091 and the boundary forcing is taken from year ~186 of the ACCESS-OM2-01 spinup. Do you still have the old scripts for which dataset you derived the boundary forcing for those runs? If not, ping me and I will figure out a good run to use. If we don't have one, I can run for another year with daily output.

**Comment by angus-g** [comment]
I do have the scripts (they're in https://github.com/COSIMA/mom6-panan), but I thought there was something about the input data disappearing — but it's just that the non-bit-groomed output isn't available. I don't think that's an issue, so I can go ahead and cut out the forcing.

**Comment by ChrisC28** [comment]
Hi all, 
Thanks for getting this started. 
As a first step, RYF (both surface and boundary) is fine. I'm completely agonstic about the year to use. However, I'd like to play around with the conditions myself at some point. 
I'm assuming the scripts in https://github.com/COSIMA/mom6-panan/scripts are where I can see how this is done?

**Comment by angus-g** [comment]
That's right. I've got a bit of a description of the process in this document too:
[MOM6 Panantarctic.md](https://github.com/COSIMA/mom6-eac/files/8075973/MOM6.Panantarctic.md)

**Comment by ChrisC28** [comment]
Do you make use of the Raphael Dussin's "brushcutter" stuff? I ask only because he has a number of example, although nothing seems to have been updated since 2017.

**Comment by angus-g** [comment]
I originally tried using it, but it wasn't suitable for our input data, and I think maybe the interpolation didn't work for scaling up to 1/20°

**Comment by ChrisC28** [comment]
Hi all, 
Has there been any progress? Anything I can do to help out at this stage?

**Comment by AndyHoggANU** [comment]
We discussed this last week - I think the consensus was that @angus-g  would shortly get into his code and produce some open BCs for us ... but I don't think we have moved too far on it yet. Angus, let us know ...

**Comment by angus-g** [comment]
This has taken me a little longer than anticipated: my previous scripts didn't scale up too well to 4 boundaries, and there was a lot of manual data massaging for panan that I didn't reincorporate to the scripts. That's all fixed now, and I have a set of boundaries and a configuration generated, which I'll commit soon. At the moment I'm running into a too-high SSH (at 294 locations):

```
WARNING from PE    52: Extreme surface sfc_state detected: i=  74 j= 306 lon=-208.650 lat= -23.7  
40 x=-208.650 y= -23.740 D= 1.0433E+01 SSH= 2.0153E+01 SST= 2.4220E+01 SSS= 9.8536E-03 U-= 0.000  
0E+00 U+= 7.1118E-01 V-= 0.0000E+00 V+= 3.7707E-02
```

They seem to be located on the coast, but I can't see any inlets or anything that were causing problems for us before.

**Comment by ChrisC28** [comment]
Thanks for getting this running. Seems odd that the dodgy SSH values are on the coast given the large-ish distance to the boundaries. I can't think of a reason straight away why that would be the case in a regional model and not in a global model.

**Comment by ChrisC28** [comment]
Russ and I just had a quick chat about this issue. He thinks that the blow-up is likely occuring in the first couple of time-steps at a single location then propagating along the coast (a "nonsense Kelvin wave"). Any chance you could run the model forward for 20 time steps, producing SSH at every time step? Could be a way to identify where the problem is first occuring.

**Comment by angus-g** [comment]
You two and Andy are on the same wavelength! Here's the first pass at the SSH, although it seems to only be hourly even though I've asked for it every timestep (currently DT = 300). Pretty clearly points to the region where the extreme values are being detected!

![zos_series](https://user-images.githubusercontent.com/11085886/158522100-1715e064-3b7c-4662-821c-a6852bfb3ea9.png)

**Comment by ChrisC28** [comment]
Where are you running these (ie. path to output)?

**Comment by angus-g** [comment]
Outputs currently at `/scratch/x77/ahg157/mom6/work/eac-01/`

**Comment by AndyHoggANU** [comment]
you might need to be on x77 to see it. If yes, just submit the request.

**Comment by ChrisC28** [comment]
Request sent

**Comment by AndyHoggANU** [comment]
Added both of you

**Comment by aekiss** [comment]
doesn't look much like a Kelvin wave to me...

**Comment by ChrisC28** [comment]
![Ringing](https://user-images.githubusercontent.com/7627726/158529982-0fa1a3c1-500c-430b-92d3-e6800227c76e.png)

**Comment by ChrisC28** [comment]
Looks kinda like ringing (3rd time step)

**Comment by ChrisC28** [comment]
If you've ever wondered what the normal modes of the system look like....

**Comment by russfiedler** [comment]
@angus-g  

Ok, looking at your ICs  I see that you have values of `-4.675` on land for your AVE_SSH and zero elsewhere. Make sure this hasn't snuck onto the wet points by mistake. Also I see you're starting from rest. I agree with Chis that this just looks like ringing at the start of the run and the time step needs to be lowered untill you establish a reasonable SSH and velocity fields. I've seen reversals of the Leeuwin and all other strange things due to the SSH heave at the start of runs.

**Comment by ChrisC28** [comment]
I was under the impression that the ICs were taken directly from the global model? Ie. shouldn't be from rest but close to balanced (at least locally)

**Comment by AndyHoggANU** [comment]
Maybe that is the problem -- we might need initial \eta, as well as T and S, (can keep u,v=0) to start up the model? My impression is that the initial state only has T and S for now??

**Comment by angus-g** [comment]
That's right, usually I've been able to spin up with just T and S without issue. I've added initial η, but the same blow-up occurs with or without initial velocities! The SSH magnitude that develops even in the first timestep swamps that from the initial state anyway.

**Comment by ChrisC28** [comment]
Looks like the baroclinic timestep is down to 50s but the tracer/thermodynamic timestep is 300s. Could that be an issue?

**Comment by angus-g** [comment]
`DT_THERM` is commented out there: they're both 50s (`MOM_parameter_doc.all` is probably the best reference for the parameters the model actually used)

**Comment by ChrisC28** [comment]
![SSH_blow_up](https://user-images.githubusercontent.com/7627726/159192860-2b2c13ec-b6fb-45dc-bc33-1b5ab7750952.png)
SSH increase is linear in time just off the QLD coast

**Comment by ChrisC28** [comment]
In fact, it seems like most of the domain shows some linear growth in the early part of the run (exception are points near the open boundaries). Eventually, the growth saturates/asymptotes, but only after eta reaches ~O(several metres).
Here's an example from mid-way between Tas and NZ. Seems to show linear growth for the first hour or so.

![Linear_growth_ex](https://user-images.githubusercontent.com/7627726/159193181-39a763bd-9be8-4922-a380-6c36ce7c922f.png)

**Comment by ChrisC28** [comment]
The effect is reduced in the south of the domain when compared with the north, but is still present. 
At first glance, it doesn't appear to be "coming from" anywhere. Just a steady, linear-ish increase in eta across the whole domain (save points near the open boundaries).

**Comment by ChrisC28** [comment]
Compare the following. Snapshot from timestep=1 and timestep=50 (ie after 2500 seconds, ~40minutes). Colourscale limits are -5 to 5m.

**Comment by ChrisC28** [comment]
![MOM6_EAC_ts_1](https://user-images.githubusercontent.com/7627726/159193459-20f4257f-d40f-4e54-9877-ed20adfc6456.png)
![MOM__EAC_ts_50](https://user-images.githubusercontent.com/7627726/159193464-6c77043e-db99-4669-b0eb-cf1f0f837140.png)

**Comment by angus-g** [comment]
I think this should be good now, with https://github.com/COSIMA/mom6-eac/commit/5eab9bcd06ab58944b3e87e667b1d22fbb0a8096

**Comment by ChrisC28** [comment]
Excellent. For the initiated, what has been changed? It's a little tricky to read the config files and to understand why the lack of a run-off field might cause blow up.

**Comment by angus-g** [comment]
@ChrisC28 summarised here: https://github.com/COSIMA/mom6-eac/issues/2#issuecomment-1076943640

I should note that I haven't run beyond the initial few days to make sure we're not hitting the same issue, time to put a proper run on the queue and see what goes wrong next!

**Comment by AndyHoggANU** [comment]
OK, so. Next question is -- who wants to run this? @ChrisC28  :: I am happy to have a crack and run for a couple of years. This might be useful as I've just been through this exercise with panant, so I might be able to spot errors more easily. But also happy for you to do the first proper run and keep us up to date. Let me know what suits best.

**Comment by ChrisC28** [comment]
For reference, the only river in eastern Australia with a signifcant run-off is the Burdekin in central north queensland, and then only in the Moonsoon period. So, this is clearly a numerical issue, no?

**Comment by angus-g** [comment]
Not even numerical, just unexpected default behaviour:
>because there is no runoff, it's using the scaling factor, 1, as a constant value! So globally, the runoff was 1 kg/m2/s over the entire domain!

**Comment by russfiedler** [comment]
@ChrisC28  for the runoff the last entry in the `data_table` can serve two functions. If the field and file exists it acts as a scaling factor. If it doesn't exist it acts as a constant field replacement variable (pretty common for mslp or if running an idealised model). In this case the file didn't exist and the values was `1` which is `1 kg m-2 s-1` or `1e-3 m s-1` or... wait for it `86 m day-1` which matches your early linear growth estimate. Pretty impressive that the model ran for so long!

**Comment by ChrisC28** [comment]
I'd be happy for @AndyHoggANU  to run for a few years just to catch anything seriously problematic going on. However, I'd like to port the model over the Pawsey system in the near future. For that, I might need some help with compiling and running (although I don't see too many problems, as I run MOM5 as part of the CSIRO coupled model on Pawsey with no issues). 

Pawsey is Cray architecture and uses slurm as a workload manager. This (obviously) breaks the payu connection, which is sub-optimal but not a huge deal.

**Comment by ChrisC28** [comment]
Thanks @russfiedler that makes perfect sense.

**Comment by ChrisC28** [comment]
It probably makes sense to run a few years on a well known system (ie. Gadi) before porting to Pawsey.

**Comment by AndyHoggANU** [comment]
I will have a crack at it and let you know how it goes.

**Comment by aidanheerdegen** [comment]
> Pawsey is Cray architecture and uses slurm as a workload manager. This (obviously) breaks the payu connection, which is sub-optimal but not a huge deal.

There was some enthusiasm for porting to other systems/schedulers see

https://github.com/payu-org/payu/issues/182

https://github.com/payu-org/payu/issues/258

but there was no compelling use case. 

This could be the compelling use case.

I know @marshallward did get it running at GFDL using `slurm`.

**Comment by ChrisC28** [comment]
There's a bit of CSIRO climate research happening on Pawsey at the moment. Richard Mataer and Dougie Squire are both running models on it at present, as well as some of the group that run in the unstructured mesh-model based on MPAS. The current HPC machine (magnus) is currently being slowly decommisioned in favour of a new HPC machine Setonix: https://pawsey.org.au/about-us/capital-refresh/ We have it on good authority that it will have similar specs to Gadi. 

porting to Pawsey/setonix could be really useful in the long term and I'd be happy to support this MOM6_EAC project being a guinea pig. Let me know what/if you need anything to support that.

**Comment by marshallward** [comment]
I was able to get Payu working on GFDL's machine which is also a Cray environment with slurm, though if I recall there was some very kludgy stuff.

[But we did abstract out a lot of PBS-specific stuff](https://github.com/payu-org/payu/tree/master/payu/schedulers) and I think with a little more work we could get it running on other Slurm-based machines.

I would be up for working on this again.  But we can speak more of this [over at the payu repo](https://github.com/payu-org/payu).

**Comment by ChrisC28** [comment]
I've opened an issue over on the Payu repo #https://github.com/payu-org/payu/issues/323

**Comment by AndyHoggANU** [comment]
Here is a snapshot of surface speed over the first 6 months. Just to say -- we have problems at the corners. Will stop for now, but let's pick this up next week!

https://user-images.githubusercontent.com/26753100/160056130-9a4d5081-49a5-43ce-bd5b-19722c8d22b8.mov

**Comment by ChrisC28** [comment]
Doesn't look too bad if you ignor the corners!

**Comment by AndyHoggANU** [comment]
In addition to the corner problem seen above, I did see one more issue at the very end of year 1:

```
FATAL from PE    59: time_interp_external 2: time 726714 (19911231.001500 is after range of list 726350-726714(19910101.000000 - 19911231.000000),file=INPUT/forcing/forcing_obc_segment_003.nc,field=u_segment_003
```

Does this mean that it is missing the last day of the boundary forcing file?

**Comment by angus-g** [comment]
I think that's right, it's hard to figure out what FMS wants! I'll append another day to the forcing tomorrow.

**Comment by angus-g** [comment]
Alright, this should be fixed. The issue wasn't a lack of days, but not setting the `modulo` attribute on the time axis of the boundary forcing. This is also why I had to hack at the units to start at `00:00:00` instead of matching `12:00:00` from the source data. So, there are two modifications to the boundary forcing:

1. time units reset to `days since 1900-01-01 12:00:00` (forcing won't be offset by 12 hours)
2. `modulo` unit attribute added to allow wrapping the year of forcing

I've also defaulted to the release, rather than debug executable in https://github.com/COSIMA/mom6-eac/commit/9a426831c82b30d9021d0da3cce4440dca2b4ff4, which might give a little speedup.

**Comment by AndyHoggANU** [comment]
OK, so should I restart from scratch?

I altered the executable in my run script, so that was all good.

Any ideas on how we diagnose the corner issues?

**Comment by angus-g** [comment]
I suppose to get the cleanest run, it should be restarted from scratch, given the boundary forcing has shifted by 12 hours. It would be interesting to see if the corners change at all in light of that. Otherwise, I'm not sure about the corners just yet!

**Comment by ChrisC28** [comment]
Regarding the corner issue, I had a quick look at the obc forcing files, thinking that maybe there could be an obvious interpolation error (or something), but now I'm confused. 
The OBC forcing files (forcing_obc_segment_00[1-4].nc) have dimension lengths that differ from those of the interior grid in a way that doesn't make a lot of sense to me. 
For example, ny_segment_004 has a length of 981, while the yt_ocean only has a length of 490. 
Are the OBC forcing files global (or large, or something) and then we let FMS truncate them to the appropriate size? I'm missing something obvious here, but it would help me to understand how the system runs.

**Comment by angus-g** [comment]
The OBCs are on the "super grid", so there's data on both the cell
corners and centres in a given direction.

**Comment by AndyHoggANU** [comment]
OK, I have started a new run. It is going smoothly as before and is 3 months in. Below you see the surface speed for day 1 and day 90. The issue seems to be right at the corner in the NE and SW. The NW corner is land, no problems there. The SE corner doesn't have a problem at all. Note that the scale here is such that red is 6m/s and the EAC region is ~1.2 m/s
<img width="164" alt="Screen Shot 2022-03-28 at 2 47 27 pm" src="https://user-images.githubusercontent.com/26753100/160323397-33f65078-c5a0-4a6b-b276-3e61052de199.png">
<img width="164" alt="Screen Shot 2022-03-28 at 2 47 45 pm" src="https://user-images.githubusercontent.com/26753100/160323473-81c1551d-95c6-423f-9d3d-7f9dde4fece6.png">

**Comment by AndyHoggANU** [comment]
Here is surface salinity from day 1. It would seem that salinity at the corners is close to zero (blue, red is 35). So I think the salinity interpolation may be creating problems...
<img width="645" alt="Screen Shot 2022-03-28 at 2 53 02 pm" src="https://user-images.githubusercontent.com/26753100/160323712-5e64c9e1-9bb3-4984-88ba-de1ebde8a1bf.png">

**Comment by AndyHoggANU** [comment]
Note that the corners are not the only problems here. What does salinity look like on the supergrid in those locations?

**Comment by ChrisC28** [comment]
There's clearly some grid-scale (and larger) noise on the eastern and southern bdys.

**Comment by aekiss** [comment]
is there salinity restoring at the surface?

**Comment by aekiss** [comment]
is the temperature also messed up at the boundary, or just salinity?

**Comment by AndyHoggANU** [comment]
@angus-g -- just to confirm that I can now successfully run multiple years, so that one is fixed.

**Comment by AndyHoggANU** [comment]
Had quick look at the EAC boundary problems today. Quick summary:
- The problems are not just at the corners, but also at a few other points on the eastern boundary
- I can only really see issues in salinity on the boundary (presumably velocity problems are a consequence)
- Temperature is fine.
- All the salt problems seem to be close to the surface.

Some sketchy plots here:
https://github.com/COSIMA/mom6-diagnostics/blob/main/EAC/Untitled.ipynb

**Comment by ChrisC28** [comment]
Having a look at @AndyHoggANU it seems that the boundary forcing files (ie. forcing_obc_segment_00*.nc) don't seem to have much in the way of obvious noise to large deviations in the salinity fields. However, the output from the model is messed up. I imagine that due to the hybrid vertical coordinates that the model must interpolate the boundary forcing somehow to apply it? Could this be a vertical coordinate matching problem? 
@russfiedler also recommends checking the applied boundary conditions (no-gradient? radiation? etc....)

**Comment by ChrisC28** [comment]
I've started attempting to compile MOM6 on Pawsey. I've run into a few stumbling blocks that I'm sure I can sort out with time. However.... 
I just realised that there is no compiling workflow for this case. 
To make some progress, I've been following @angus-g 's instructions here: https://github.com/angus-g/mom6-ninja-nci
But they are quite generic. Ditto the MOM6 wiki. 
Is there a clear workflow for compiling this EAC 10 degree case? Or is the whole configuration (grid, forcing, etc...) set at run time? Would some nice person (@angus-g ?) be able to point me to an example on how this case has been compiled on Gadi?
Once again, sorry for the vague question.

**Comment by angus-g** [comment]
> Or is the whole configuration (grid, forcing, etc...) set at run time?

Yep, the executable is completely generic to any MOM6-SIS2, symmetric, coupled configuration! I compiled it directly with those scripts. I can probably give you some help compiling on Pawsey if you need.

**Comment by ChrisC28** [comment]
So all the grid definitions, etc... are computed at run-time? It's a brave new world! 

 I'll have another attempt at compilation on Pawsey next week some time (a few deadlines this week).

**Comment by angus-g** [comment]
That's right, grid sizes are dynamic. It is possible, but awkward to statically-define the grid size. The only important compile-time decision is whether the grid should be symmetric or non-symmetric w.r.t. cell centres, i.e. is the rightmost/topmost point on the grid a centre or an edge. Usually symmetric is the better choice here, so that's my default.

**Comment by russfiedler** [comment]
I had a look at the river runoff in @AndyHoggANU 's run for output007. The good news is that MOM6 is very stable with a very high runoff. The bad news is that a runoff of 700m/day (that is m/day not mm/day) in the middle of the ocean might be considered somewhat unrealistic. 
![oh_dearie_me](https://user-images.githubusercontent.com/12705799/160987467-f90c77ab-834f-47dc-8eff-1e54c5681767.gif)

**Comment by angus-g** [comment]
I wish the data_table was smart enough to not just dump everything outside the domain on the edges! Seems like the runoff regridding has gone awry then...

**Comment by ChrisC28** [comment]
![Run_off_noise_eac10](https://user-images.githubusercontent.com/7627726/160988476-14aed788-2c28-4191-8bf7-fae91f4c90fc.png)
Run off strikes again! (/g/data/x77/ahg157/inputs/mom6/eac-01/eac_runoff.nc)

**Comment by angus-g** [comment]
Maybe we should just go back to 0 runoff!

**Comment by russfiedler** [comment]
@ChrisC28 so are you saying it's actually the preprocessed runoff that's the problem and not FMS?

**Comment by ChrisC28** [comment]
@russfiedler looks like it to me but I'll defer to @AndyHoggANU and @angus-g

**Comment by russfiedler** [comment]
`"ICE", "runoff", "friver", "./INPUT/eac_runoff.nc", "none", 1.0`
 Answer my own question. This line says the runoff is on the model grid.

**Comment by AndyHoggANU** [comment]
Yeah. OK. We do need to regrid the runoff, but it seems to have gone awry. 
Good catch as always @russfiedler

**Comment by AndyHoggANU** [comment]
PS. Yes, it is actually the `eac_runoff.nc` file which is adding extra runof to the boundaries and causing our problems.
@angus-g  -- should be an easy fix, right? Can you have a go at recreating that file? It could actually just be sliced out of the original, given there is no actual regridding to be done here. If you get it done tomorrow, ping me and I will try to kick off the run from scratch again.

**Comment by aekiss** [comment]
It'd be nice to be able to automate this somehow. 

Would it work to generate `eac_runoff.nc` on a grid that is 1 row bigger than we need on each edge, then slice 1 row off each of the 4 edges? Land edges will be unaffected (runoff will already be at the coast in the inner grid) but it will slice off the runoff that was mapped to the open ocean edges.

**Comment by russfiedler** [comment]
I'm not sure why you would want to do that if you can cookie cutter it to the model T points and avoid any further interpolation without checking location of points. The problem appears to be the naive application of a nearest neighbour interpolation program which was intended for a global grid.
I suppose you could make an intermediate file and then cut out the required subset.
Arrgh strike that last line. That's what you are suggesting. Yes it should work.

**Comment by ChrisC28** [comment]
It seems odd to me that nearest-neighbour won't interpolate run-off into ocean points far from land?

**Comment by russfiedler** [comment]
@ChrisC28 It depends if the algorithm is searching for the closest coastal point or the closest wet point. It must be the closest wet point given what we see but there might be some other logic with boundaries.

**Comment by AndyHoggANU** [comment]
OK, well this is looking a lot better. See /g/data/ik11/outputs/mom6-eac/eac-zstar-v2 for details.

Note that there is still some funniness in some corners near PNG, but it is a good start. Time to think about next steps?

https://user-images.githubusercontent.com/26753100/162091870-58ee1422-18f3-44e1-bf47-6d467bde1ecc.mov

**Comment by ChrisC28** [comment]
Yeah, run-off still seems a bit high in certain regions. In particularly, the region to the west of bass-strait/Tasmania and along the southern boundary still has some issues. For example, have a look (carefully) at this image from /scratch/x77/amh157/mom6/work/mom6-eac/INPUT/runoff/eac_runoff.nc

![eac_runoff_new](https://user-images.githubusercontent.com/7627726/162098759-b84327c7-30c2-467a-b017-ced325202792.png)

**Comment by angus-g** [comment]
The runoff file being used as of #4 is `runoff_regrid.nc`

**Comment by ChrisC28** [comment]
Whopps, should have checked before poking around.

**Comment by ChrisC28** [comment]
Next steps: I'd like to get this config running on Pawsey. Still need to get the model compiled. Any takers to help me get the model compiled? Bit snowed under today, but may have some time tomorrow?

**Comment by aidanheerdegen** [comment]
As I'm already set up with your project I can take a look.

**Comment by aidanheerdegen** [comment]
I've created a `MOM6` executable here
```bash
/group/pawsey0410/aidanh/ocean_only_symmetric/MOM6
```

Using this branch on a fork of Angus' repo

https://github.com/coecms/mom6-ninja-nci/tree/pawsey

This was compiled using the Intel compiler because it was easy as it was set up for Intel.

Next step is to test with a simple MOM6 example.

**Comment by ChrisC28** [comment]
Awesome. The other immidate next steps that I have on my "wish list" #3:
- high (1/20th or 1/50th deg) and low (1 degree) resolution configurations;
- interannually forced runs.

**Comment by aidanheerdegen** [comment]
Successfully run `double_gyre` test in `/group/pawsey0410/aidanh/experiments/double_gyre`.

I'll look into running the same but with `payu` next.

**Comment by aidanheerdegen** [comment]
I have made a `conda` environment with an updated `payu` from this branch/PR

https://github.com/payu-org/payu/pull/326

To use add this to your `~/.bashrc`
```bash
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/group/pawsey0410/mambaforge/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/group/pawsey0410/mambaforge/etc/profile.d/conda.sh" ]; then
        . "/group/pawsey0410/mambaforge/etc/profile.d/conda.sh"
    else
        export PATH="/group/pawsey0410/mambaforge/bin:$PATH"
    fi
fi
unset __conda_setup

if [ -f "/group/pawsey0410/mambaforge/etc/profile.d/mamba.sh" ]; then
    . "/group/pawsey0410/mambaforge/etc/profile.d/mamba.sh"
fi
# <<< conda initialize <<<
```
and then `conda activate payu`.

Then `payu` should "just work" with a few modifications to `config.yaml`.

I have an updated `mom6_double_gyre` config in this branch

https://github.com/aidanheerdegen/mom6_double_gyre/tree/pawsey

which is located here on magnus `/group/pawsey0410/aidanh/experiments/payu/mom6_double_gyre`

Looking at the changes compared to `gadi` helps to show what is required:
```diff
diff --git a/config.yaml b/config.yaml
index 58acbb8..bd9b354 100644
--- a/config.yaml
+++ b/config.yaml
@@ -1,23 +1,26 @@
 # PBS flags
-queue: express
-project: x77
+# queue: express
+# project: x77
+scheduler: slurm
 walltime: 0:30:00
-ncpus: 2
-mem: 4GB
+ncpus: 8
+# mem: 4GB
 jobname: double_gyre
 runspersub: 15
 
 collate: false
-collate_walltime: 0:30:00
-collate_mem: 30GB
+
 # Model config
 model: mom6
+
 # model: gold
-shortpath: /short/v45
-exe: MOM6
+shortpath: /group/pawsey0410
+exe: /group/pawsey0410/aidanh/mom6-ninja-nci/ocean_only_symmetric/MOM6
 input: double_gyre
 priority: 15
 stacksize: unlimited 
 
 runlog: true
 
+mpi:
+  runcmd: srun
```

Let me know if you need any assistance, or if you want me to get the EAC config up and running.

**Comment by ChrisC28** [comment]
Attempted to run the double gyre case following @aidanheerdegen instructions above and it seems I'm not configuring one of the model or payu correctly. After cloning the mom6_double_gyre case, and copying the config.yaml file above, when I run `payu setup ` I get the following error: 
`payu: error: Input directory /group/pawsey0410/cc7576/mom6/input/double_gyre not found; aborting.`

I'm attempting to run this from `/group/pawsey0410/cc7576/MOM6/experiments/mom6_double_gyre`
Likely an easy fix (I'm probably doing something silly) but beyond my payu knowledge. Any ideas?

**Comment by aidanheerdegen** [comment]
Sorry Chris, my instructions were insufficient.

You'll need to do 
```
payu init
```
to create the `laboratory` structure and add the files from `/group/pawsey0410/aidanh/mom6/input/double_gyre/` to `/group/pawsey0410/cc7576/mom6/input/double_gyre/`, or change the line
```yaml
input: double_gyre
```
to
```yaml
input: /group/pawsey0410/aidanh/mom6/input/double_gyre/
```
and just pick up the input files from my laboratory. The latter is the easiest in the first instance.

**Comment by ChrisC28** [comment]
Does it matter where `payu init` is run from? Will that affect the laboratory structure?

**Comment by ChrisC28** [comment]
Answered my own question above: no, it doesn't matter where `payu init` is run.

**Comment by aidanheerdegen** [comment]
It has to be run in the `control` directory (where `config.yaml` resides), or in any similar `mom6` control directory.

It is making a directory structure like this:
```bash
/group/pawsey0410/aidanh/mom6/archive
/group/pawsey0410/aidanh/mom6/input
/group/pawsey0410/aidanh/mom6/work
/group/pawsey0410/aidanh/mom6/codebase
/group/pawsey0410/aidanh/mom6/bin
```
but if you just pointed to my `input` dir it wouldn't complain, and would make that directory structure for you once it needed to archive stuff (or as much of as it as is required, can't quite recall details).

**Comment by ChrisC28** [comment]
Good news is that payu seems to be behaving. Bad news is that `mom6` is now crashing: 
```
FATAL from PE     3: MOM_domains_init: The product of the two components of layout,    1   2, is not the number of PEs used,     8.
Rank 3 [Mon Apr 11 08:26:57 2022] [c0-0c2s3n0] application called MPI_Abort(MPI_COMM_WORLD, 1) - process 3
```

**Comment by ChrisC28** [comment]
config file is simply copied from the above with "ncpus: 8"

**Comment by aidanheerdegen** [comment]
Sorry again, I also made the following change to the `MOM_input` file not reflected in that experiment repo:
```diff
ndex 0b3d0c6..2e4d5ae 100644
--- a/MOM_input
+++ b/MOM_input
@@ -49,8 +49,8 @@ NJGLOBAL = 40                   !
 !NJPROC = 4                     !
                                 ! The number of processors in the x-direction. With
                                 ! STATIC_MEMORY_ this is set in MOM_memory.h at compile time.
-LAYOUT = 1, 2                   !
-                                ! The processor layout that was acutally used.
+! LAYOUT = 1, 2                   !
+!                                 ! The processor layout that was acutally used.
 IO_LAYOUT = 1, 1                ! default = 0
                                 ! The processor layout to be used, or 0,0 to automatically
                                 ! set the io_layout to be the same as the layout.
```
So I commented out the `LAYOUT` config option.

**Comment by aidanheerdegen** [comment]
To run multiple times, say 10, use 
```
payu run -n 10
```
You'll note that I've set 
```yaml
runspersub: 15
```
which means it will run 15 times in a single slurm submit. It is for testing, but works well to reduce queue overhead for experiments that run relatively quickly, but which for other reasons you don't want to run longer for each submit.

`payu` @ pawsey is still a little rough around the edges. I have just done enough to get it working. More may still be required for that, but proper support requires a re-write of some of the internal plumbing/assumptions.

**Comment by ChrisC28** [comment]
Thanks @aidanheerdegen I appear to have got the double_gyre case working. 
Regarding payu @pawsey, it seems to be working "well enough" for now. Only thing I'd add is that it would be useful to put the `work` directory on `\scratch` for more realistic runs. We can deal with additional support for Pawsey when and if it arises and I may be able to get some support from CSIRO for that. 
I'm going to make an attempt to get `eac_10` working today. Almost certain I'll run into an issue I can't solve, so, thanks in advance for the continuing help! It's been quite a while (~9 years) since I used payu or MOM.

**Comment by aidanheerdegen** [comment]
I should have put the `payu` stuff in https://github.com/payu-org/payu/issues/323

Will reply there.

**Comment by ChrisC28** [comment]
Attempted to run mom6-eac-10 on Pawsey today. Ran into my first stumbling block:
`FATAL from PE    58: fms_io(read_data_3d_new), field temp in file INPUT/forcing/init_cond.nc: field size mismatch 1
`
I've copied the input, forcing and initial conditions from Gadi. In particular, the offending file `init_cond.nc` has been copied from 
`/g/data/x77/ahg157/inputs/mom6/eac-01`

On Pawsey, I've put the eac specific inputs: 
 `/group/pawsey0410/cc7576/mom6/input/eac-01/`
and the JRA forcing: 
`/group/pawsey0410/cc7576/mom6/input/JRA55-do/RYF/v1-3/
`
Does this error look familiar to anyone?

**Comment by angus-g** [comment]
I believe you're missing a patch for FMS2 field sizes, the executable we're using on gadi is built from this branch: https://github.com/angus-g/mom6/tree/fms2-uncompress-restart

**Comment by ChrisC28** [comment]
I'm using @aidanheerdegen build on Pawsey: `exe: /group/pawsey0410/aidanh/mom6-ninja-nci/ocean_only_symmetric/MOM6
`

**Comment by ChrisC28** [comment]
He'll have a better idea of the provenance of the executable.

**Comment by aidanheerdegen** [comment]
I have switched to @angus-g 's `MOM6` branch here:
```
/group/pawsey0410/aidanh/MOM6-examples/src/MOM6
```
like so:
```
git remote add angus git@github.com:angus-g/MOM6.git
get checkout angus/fms2-uncompress-restart -b fms2-uncompress-restart
```
and then recompiled here `/group/pawsey0410/aidanh/mom6-ninja-nci` by running `./cms_build.sh `

So `/group/pawsey0410/aidanh/mom6-ninja-nci/ocean_only_symmetric/MOM6` should be updated.

NB: I haven't compiled for the target architecture of the compute nodes on magnus. As a result the code *might* not be running as as it could, but IIRC the various vectorisation flags didn't make much difference on `gadi`.

**Comment by angus-g** [comment]
Oh, I've also just noticed that you're using the ocean only executable. For the JRA forcing, you'll need the coupled ice-ocean executable.

**Comment by aidanheerdegen** [comment]
*grumble*

**Comment by aidanheerdegen** [comment]
SIS2?

**Comment by angus-g** [comment]
Yep, the `ice_ocean_SIS2_symmetric` configuration is the one. Here's the manifest:
```yaml
- exe: "symmetric_FMS2-e7d09b7"
  build-date: 2022-01-20
  sha256: "e7d09b7687fba4cf0693adf3c1a5f653dee312df996b99da384c64c279a53eef"
  git:
    - component: MOM6
      ref: "93968bd8ec1a993c705846ec91c4cf33b86bc4fb"
    - component: SIS2
      ref: "d18605eacdd2c92b8262458fcad107c546dd080f"
    - component: FMS
      ref: "2d373d20af28b9d8e2cc0d0acf2334fcff457e47"
  modules:
    - "intel-mkl/2020.2.254"
    - "python3/3.8.5"
    - "netcdf/4.7.4p"
    - "openmpi/4.1.2"
    - "intel-compiler/2021.4.0"
  keywords: [SIS2, symmetric, FMS2]
  flags:
    fflags: "-fno-alias -auto -safe-cray-ptr -ftz -assume byterecl -i4 -r8 -nowarn -sox -g -O2 -debug minimal -fp-model precise -qoverride-limits -xHost"
    cflags: "-D__IFC -sox -g"
```

**Comment by aidanheerdegen** [comment]
@ChrisC28 change the `exe` to point here
```
/group/pawsey0410/aidanh/mom6-ninja-nci/ice_ocean_SIS2_symmetric/MOM6
```
and give it a burl

**Comment by ChrisC28** [comment]
Ok. Need to use a coupled ice-ocean model in the mid-latitudes. Makes sense. 

Thanks again. I'll take a stab at it.

**Comment by angus-g** [comment]
It is possible to force the ocean-only model from files (https://github.com/mom-ocean/MOM6/blob/399a7db1565e89e0ac8aeedd2de32bc846b1f230/config_src/drivers/solo_driver/MOM_surface_forcing.F90#L1594-L1647), but I haven't set that up for JRA55. I'm sure you can just wind the ice timestep up to the coupling (i.e. forcing) timestep and it won't be much of an expense. With the defaults in this configuration, the ice is maybe 3% or so of the overall runtime.

**Comment by ChrisC28** [comment]
Hit another stumbling block. Fails on the first timestep and appears to incorrectly initialize: 
The model is falling over in a function contained in the `regrid_interp.F90` file `get_polynomial_coordinate` which is part of the ALE stuff. The error is: 
```
MOM Date   1991/01/01 00:00:00      0: En 2.017888E-03, MaxCFL  0.10825, Mass 3.676899176519E+19, Salt  34.70639731170, Temp   4.85064905983
    Total Energy: 43707989AC11AD58  7.4195709307049344E+16
    Total Mass:   3.6768991765185659E+19, Change:   0.0000000000000000E+00 Error:  0.00000E+00 ( 0.0E+00)
    Total Salt:   1.2761192369530243E+18, Change:   0.0000000000000000E+00 Error:  0.00000E+00 ( 0.0E+00)
    Total Heat:   7.1198707354443406E+23, Change:   0.0000000000000000E+00 Error:  0.00000E+00 ( 0.0E+00)

WARNING from PE     0: btstep: eta has dropped below bathyT:  -1.8214812726310928E+65 vs.  -4.4050000000000000E+03 at  -2.1585E+02 -4.8061E+01      3      2


WARNING from PE     0: btstep: eta has dropped below bathyT:  -1.7707721182959732E+65 vs.  -4.4683457031250000E+03 at  -2.1575E+02 -4.8061E+01      4      2


FATAL from PE     0:  Could not find target coordinate   1033.00000000000      in get_polynomial_coordinate. This is caused by an inconsistent interpolant (perhaps not monotonically increasing):   1033.00000000000                          NaN                     NaN
```
It's possible (likely) that I haven't configured the vertical coordinate correctly.

**Comment by angus-g** [comment]
I think that regridding failure is a red herring: the sea level being in a different galaxy is probably the root cause! Maybe some missing/incorrect forcing, or the initial state is wrong?

**Comment by ChrisC28** [comment]
Agreed, although I wasn't sure of the chain of causality. I've taken the forcing/initial conditions from: `/g/data/x77/ahg157/inputs/mom6/eac-01/`
and checked the `MOM_input` file to make sure that it's equivalent to the one run by @AndyHoggANU on gadi (ie. `/scratch/x77/amh157/mom6/archive/mom6-eac/output001/MOM_input`). 
I'll have another poke around a little later, but missing forcing is a possible culprit.

**Comment by angus-g** [comment]
Have you got all the JRA55 forcing for 1990-1991 from `/g/data/ik11/inputs/JRA-55/RYF/v1-3`?

**Comment by ChrisC28** [comment]
No. I'm using`: /g/data/ua8/JRA55-do/RYF/v1-3`

**Comment by angus-g** [comment]
That's alright, they're the same.

**Comment by ChrisC28** [comment]
Still haven't found the source of this error and Im clearly missing a step or two. 
Essentially my workflow has been (thus far):
- copy the forcing from Gadi to Magnus (Pawsey);
- clone the mom6-eac repo 
- update the config.yaml file by mashing together the working config.yaml for the double gyre test case and the config.yaml from the 
- run `payu setup` and `payu run`
Clearly the model is not initialising correctly, but I can't track down why. 

Can some wonderful person (maybe @AndyHoggANU ) point me to a the directories on Gadi that were used to run the model, so I can check the differences between what's on Pawsey and what works?

**Comment by ChrisC28** [comment]
Also, my `config.yaml` file is below. Please let me know if there is anything obviously wrong: 
```

scheduler: slurm
walltime: 0:61:00
ncpus: 91
# mem: 4GB
jobname: mom6_eac_01
runspersub: 1


# Model config
model: mom6

# model: gold
shortpath: /group/pawsey0410
exe: /group/pawsey0410/aidanh/mom6-ninja-nci/ice_ocean_SIS2_symmetric/MOM6
stacksize: unlimited

input:
    - /group/pawsey0410/cc7576/MOM6/inputs/mom6/eac-01
    - /group/pawsey0410/cc7576/JRA55-do/RYF/v1-3

collate: false
runlog: true

mpi:
  runcmd: srun
~                                                                                                                                                                                                                                                                                  
~             
```

**Comment by aidanheerdegen** [comment]
I can take a look @ChrisC28 

Where is your control directory (where you type `payu run`)

**Comment by ChrisC28** [comment]
Thanks @aidanheerdegen 
control dir: `/group/pawsey0410/cc7576/MOM6/experiments/mom6-eac`

**Comment by ChrisC28** [comment]
It _should_ look somewhat like you would expect.

**Comment by aidanheerdegen** [comment]
I have recompiled the binary with traceback flags to nail down where the error is occurring. This is the traceback
```
forrtl: error (76): Abort trap signal
Image              PC                Routine            Line        Source             
MOM6               000000000207777A  Unknown               Unknown  Unknown
MOM6               0000000001B61740  Unknown               Unknown  Unknown
MOM6               000000000205AD5B  Unknown               Unknown  Unknown
MOM6               000000000215E9FA  Unknown               Unknown  Unknown
MOM6               0000000001EA66F2  Unknown               Unknown  Unknown
MOM6               0000000001E864CB  Unknown               Unknown  Unknown
MOM6               0000000001E87995  Unknown               Unknown  Unknown
MOM6               00000000015972A1  mpp_mod_mp_mpp_er          69  mpp_util_mpi.inc
MOM6               00000000005B35EC  regrid_interp_mp_         429  regrid_interp.F90
MOM6               00000000005B7D82  coord_rho_mp_buil         149  coord_rho.F90
MOM6               00000000004902A9  mom_diag_remap_mp         330  MOM_diag_remap.F90
MOM6               00000000005352CF  mom_diag_mediator        3410  MOM_diag_mediator.F90
MOM6               0000000000A5F00C  mom_mp_step_mom_         1273  MOM.F90
MOM6               0000000000F49A15  ocean_model_mod_m         584  ocean_model_MOM.F90
MOM6               00000000012263D5  MAIN__                   1082  coupler_main.F90
MOM6               000000000040A29E  Unknown               Unknown  Unknown
MOM6               0000000002158779  Unknown               Unknown  Unknown
MOM6               000000000040A189  Unknown               Unknown  Unknown
```
Given there is a `NaN` in the error message
```
FATAL from PE    26:  Could not find target coordinate   1033.00000000000      in get_polynomial_coordinate. This is caused by an inconsistent interpolant (perhaps not monotonically increasing):   1033.00000000000        1007.80205022543                          NaN
```
and the relevant code is:
```fortran
  if ( k_found == -1 ) then
    write(mesg,*) 'Could not find target coordinate', target_value, 'in get_polynomial_coordinate. This is '//&
                  'caused by an inconsistent interpolant (perhaps not monotonically increasing):', &
                  target_value, edge_values(1,1), edge_values(N,2)
    call MOM_error( FATAL, mesg )
  endif
```
It suggests where this is being called (`coord_rho.F90:432`) is passing in some nonsense values for density
```fortran
    ! Based on source column density profile, interpolate to generate a new grid
    call build_and_interpolate_grid(CS%interp_CS, densities, count_nonzero_layers, &
                                    h_nv, xTmp, CS%target_density, CS%nk, h_new, &
                                    x1, h_neglect, h_neglect_edge)
```

**Comment by aidanheerdegen** [comment]
(hit the wrong button, apologies)

**Comment by russfiedler** [comment]
I think you either (1) Need to trap NaNs or (2) use some of the new compile switches to initialise values to non NaN. It's quite possible that a multiplication mask is being used at some stage and NaNs won't be masked away. I'd suggest an initial value like HUGE. Also, are you using the same tile mask and layout as was used on Gadi?

**Comment by ChrisC28** [comment]
Hmmmm.... mask and vertical coordinate files are taken directly from the example on Gadi (I simply copied over the directory `/g/data/x77/ahg157/inputs/mom6/eac-01/` ). 
Is this an issue with the intel compiler on Pawsey? Seems odd that it would handle NaNs on Gadi but not Pawsey.

**Comment by ChrisC28** [comment]
Just to check my workflow, I attempted to run on Gadi today. I ran it no problems. The only real change between the Pawsey and Gadi configs (beside the HPC system) is the config.yaml file. 

I'm now 99% convinced this is a Pawsey specific error (if I wasn't before).

**Comment by ChrisC28** [comment]
Hi all, 
@angus-g has sorted out whatever the issue was. Unsure exactly what was making the model fall over, but likely my error. 

I'm going to try to run a few years on pawsey and then start hasselling you all about changing resolution, etc...

**Comment by AndyHoggANU** [comment]
Awesome. Just an idea for early on -- it would be good to compare model cost between Pawsey and NCI for this config. Maybe @angus-g could take a look at that in his spare time? Doesn't need a long run or detailed profiling, just nice to have an idea on whether Pawsey is a good fit (or a better option) for MOM6?

**Comment by angus-g** [comment]
Without looking too closely, the equivalent first month run between the systems is pretty different:

period  | system | initialisation | main loop   | termination
------- | ------ | -------------- | ----------- | -----------
1 month | Magnus | 11.957479      | 2793.179132 | 21.559437
1 month | Gadi   | 4.608698       | 2181.395742 | 17.397071

**Comment by AndyHoggANU** [comment]
Yeah, 36% longer on Pawsey ... but it is without any optimisation at all. Also, what is the SU cost over at Pawsey?

**Comment by angus-g** [comment]
I believe it's just 1 SU = 1 CPU hour, and the Magnus nodes are 24 CPUs. Setonix will have 128 CPUs/node, but will do proportional charging for less than a node. It's unclear to me whether you can a non-multiple of 128 CPUs once you ask for more than one node though.

**Comment by ChrisC28** [comment]
From the Pawsey training materials: "1 SU is equivalent to 1 hour use of 1 CPU core."

**Comment by ChrisC28** [comment]
What level of compiler optimisation are we using at present?

**Comment by ChrisC28** [comment]
With 128 CPUs per node, would using a hybrid MPI/OPEN-MP approach be worth investigating? I don't know if this has been done with MOM6 in the past.

**Comment by angus-g** [comment]
> What level of compiler optimisation are we using at present?

`-O2`, we might not be switching on the platform specific optimisations (some of the AVX stuff), but MOM6 doesn't vectorise particularly well so I don't know if that would make too much of an impact.

> would using a hybrid MPI/OPEN-MP approach be worth investigating?

For a while, the OpenMP directives in MOM6 were pretty actively maintained, but I don't know if anybody actually uses them. I think most MPI implementations are pretty fast for same-node operations (using shared memory, etc.), but I certainly haven't benchmarked any of this.

**Comment by aidanheerdegen** [comment]
Just on clock speed along (2.6GHz v 3.2GHz) you'd expect `gadi` to be almost 25% faster. Add in 50% faster interconnect fabric, more cores per node (so lower inter-node traffic) and the speed difference looks about right.

**Comment by ChrisC28** [comment]
Does Gadi still operate on 2SU = 1 CPU hour?

**Comment by aidanheerdegen** [comment]
For the `normal` queue yes but IIRC they doubled our allocation, so ....

**Comment by AndyHoggANU** [comment]
But, the point is -- even though Pawsey is slower, it costs less than the normal queue on Gadi.

**Comment by ChrisC28** [comment]
*If SUs are allocated in the same way (ie. if the accounting at NCI doesn't explicitly factor in the pricing of SUs)

**Comment by ChrisC28** [comment]
Over the weekend I ran the model for 5 years on Pawsey. The output looks good! Thanks for all your help getting this far. 

The next steps I'd like to persue are: 
- changing from RYF to IAF;
- change to a higher resolution. 
I imagine that the former is simpler than the later (happy to be corrected) 
Would it be better to close this issue and open another now that the intital teething issues have been worked through?

**Comment by AndyHoggANU** [comment]
Yes, let's close this issue and open up 2 new ones. I will comment more there ...
